lib/migrate.js

```diff
@@ -0,0 +1,207 @@
+    * @param {Function} cb
+  */
+  Cloudant.prototype._createIndex = function(mo, model, name, indexObj, cb) {
+    const self = this;
+    debug('createIndex fields before coercion: %s', inspect(indexObj.fields));
+    let fields = self.coerceIndexFields(indexObj.fields);
+    debug('createIndex fields after coercion: %s', inspect(fields));
+    fields = self.addModelViewToIndex(mo.modelView, fields);
+    debug('createIndex fields after add model index: %s', inspect(fields));
+    // naming convertion: '_design/LBModel__Foo__LBIndex__foo_index',
+    // here the driver api takes in the name without prefix '_design/'
+    const config = {
+      ddocName: self.getIndexModelPrefix(mo) + '__' + model + '__' +
+      self.getIndexPropertyPrefix(mo) + '__' + name,
+      indexName: name,
+      fields: fields,
+      partitioned: indexObj.partitioned,
+    };
+    self.createIndexWithConfig(config, cb);
+  };
+
+  /**
+   * Used in function `getModifyIndexes()`.
+   * Generate indexes for model properties that are configured as
+   * `{index: true}`, return model property indexes as objects,
+   * e.g. {fields: [{name: 'asc'}], partitioned: true}.
+   *
+   * @param {Object} indexes indexes from model config, retrieved in
+   * `getModifyIndexes()`
+   */
+  Cloudant.prototype._generatePropertyLevelIndexes = function(indexes) {
+    const results = {};
+    for (const key in indexes) {
+      const field = {};
+      // By default the order will be `asc`, partitioned is false,
+      // please create Model level index if you need `desc` or partitioned index
+      field[key.split('_index')[0]] = 'asc';
+      const fields = [field];
+      results[key] = {fields: fields, partitioned: false};
+    }
+    return results;
+  };
+
+  /**
+   * Used in function `getModifyIndexes()`.
+   * Generate indexes for indexes defined in the model config.
+   * Return indexes as objects, e.g.
+   * {fields: [{foo: 'asc'}, {bar: 'desc'}], partitioned: true}
+   *
+   * @param {Object} indexes indexes from model config, provided by
+   * `getModifyIndexes()`
+   */
+  Cloudant.prototype._generateModelLevelIndexes = function(indexes, cb) {
+    const results = {};
+    for (const key in indexes) {
+      const keys = indexes[key].keys;
+      if (!keys || typeof keys !== 'object') return cb(new Error(
+        'the keys in your model index are not well defined! please see' +
+      'https://loopback.io/doc/en/lb3/Model-definition-JSON-file.html#indexes',
+      ));
+
+      const partitioned = indexes[key].partitioned || false;
+
+      const fields = [];
+      _.forEach(keys, function(value, key) {
+        const obj = {};
+        let order;
+        if (keys[key] === 1) order = 'asc';
+        else order = 'desc';
+        obj[key] = order;
+        fields.push(obj);
+      });
+      results[key] = {fields, partitioned};
+    }
+    return results;
+  };
+
+  /**
+ * Perform the indexes comparison for `autoupdate`.
```

Comment:
Is this function for comparing if indexes exist?

Reply:
@agnes512 yep, it returns the new indexes to add and old indexes to drop